pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -291,70 +274,99 @@ pub fn schedule_timeline_checkpoint_upload(
     }
 }
 
+pub fn schedule_timeline_download(tenant_id: ZTenantId, timeline_id: ZTimelineId) {
+    sync_queue::push(SyncTask::new(
```

Comment:
First to note, downloads are now scheduled for CloudOnly timelines only, once per request.
There's a way to spam get_timeline_info requests or somehow overlap timeline queries to add multiple download tasks, but I have the dedup functionality for such case, so every task batch we fetch contains different tasks.

Download gathers target timeline's files, and uncompression skips such, so even if the download tasks gets to multiple batches, we'll spend ~1s for such task and reload the timeline in pageserver's memory.

As for the uploads, those are triggered by the checkpoints, which are not done for the remote only timelines.
Some code could change this, but it's a different story again 🙂 

Upload happens from local files only (otherwise, the compression would fail and the upload tasks will be evicted, never getting into the remote timeline index), so even if somehow the download happens after uploads, download will skip already present local files (well, metadata is special and will be overwritten)

Reply:
>  downloads are now scheduled for CloudOnly timelines only

This is great!

> As for the uploads, those are triggered by the checkpoints, which are not done for the remote only timelines

Timeline might be remote for one pageserver and local for another, can be local for two of them at once so this is possible :) I've created #971 to discuss that

Just to note there is a gc api created for tests but me might want to have it for something else. And it triggers checkpoint. But concurrent checkpointing is s not possible anyway within one repo, so this is not a problem


> (well, metadata is special and will be overwritten)

I think we should somehow distinguish newer/older metadata and keep the newest one. Or have some other well defined behaviour for that. I e extract metadata first and check how this checkpoint relates to timeline state (earlier or later) and decide what to do with existing files/metadata

Reply:
Nice idea in general, but not sure that needed now:
* first, uploads happen only after checkpoints, which are carried out for local timelines only (i.e. ready in the context of the LayeredRepo) 
* second, I think it would be beneficial to keep track of sync states in the index too (so we could deserialize the index and resume the sync), but that task is adjacent to the index access from the rest of the pageserver

So, a check looks good now, but it feels excessive and redundant now and might be good to add later.